### PR TITLE
trip manager final

### PR DESCRIPTION
**Pull request by obedjamir**
All functionalities implemented

### CODE REVIEW

File: core/src/main/java/com/bankbazaar/core/manager/TripActivityEntityManager.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import com.bankbazaar.core.repository.TripRepository;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.Optional;
+@Slf4j
+@Service
+public class TripActivityEntityManager {
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/core/manager/TripActivityEntityManager.java]: Manager names need not have Entity. ex TripActivityManager

File: core/src/main/java/com/bankbazaar/core/manager/TripActivityEntityManager.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    @Autowired
+    private TripRepository tripRepository;
+
+    /**
+     * Insert to TripActivity table
+     * @param data
+     */
+    public TripActivityEntity saveTripActivity(TripActivityEntity data)
+    {
+        Optional<TripEntity> tripId = tripRepository.findById(data.getTripId().getTripId());
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/core/manager/TripActivityEntityManager.java]: Dont understand this. We are fetching the trip using the id and then replacing the id

File: core/src/main/java/com/bankbazaar/core/manager/TripActivityEntityManager.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        return tripActivityRepository.save(data);
+    }
+
+    /**
+     * update to TripActivity table
+     * @param data
+     */
+    public TripActivityEntity updateTripActivity(TripActivityEntity data)
+    {
+        Optional<TripActivityEntity> presentData = exists(data.getTripId().getTripId(),data.getActivityId());
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/core/manager/TripActivityEntityManager.java]: If you have the activity id already do we need to fetch using tripId?

File: core/src/main/java/com/bankbazaar/core/manager/TripUserMapEntityManager.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    /**
+     * Insert to TripUserMapping table
+     * @param data
+     */
+    public TripUserMapEntity saveTripUserMapping(TripUserMapEntity data)
+    {
+        Optional<TripEntity> tripId = tripRepository.findById(data.getTripId().getTripId());
+        Optional<UserEntity> userId = usersRepository.findById(data.getUserId().getUserId());
+        data.setTripId(tripId.get());
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/core/manager/TripUserMapEntityManager.java]: It says tripId and userId but we are setting the entire objects

File: core/src/main/java/com/bankbazaar/core/model/TripActivityCompositeKey.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import lombok.NoArgsConstructor;
+
+import java.io.Serializable;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class TripActivityCompositeKey implements Serializable {
+    private Long tripId;
+    private Long activityId;
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/core/model/TripActivityCompositeKey.java]: Why is this required to be a composite key?

File: core/src/main/java/com/bankbazaar/core/model/TripActivityCompositeKey.java [hunk 6]
```diff
@@ -0,0 +1,16 @@
+package com.bankbazaar.core.model;
+
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/core/model/TripActivityCompositeKey.java]: @Data itself handles the constructors

File: core/src/main/java/com/bankbazaar/core/repository/TripActivityRepository.java [hunk 7]
```diff
@@ -1,7 +1,14 @@
...
 import com.bankbazaar.core.model.TripActivityEntity;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
 
-public interface TripActivityRepository extends JpaRepository<TripActivityEntity, Long> {
+import java.util.List;
+
+public interface TripActivityRepository extends JpaRepository<TripActivityEntity, TripActivityCompositeKey> {
+
+    @Query("SELECT t FROM TripActivityEntity t WHERE t.tripId.tripId = ?1")
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/core/repository/TripActivityRepository.java]: Why do we need @Query here?

File: core/src/main/java/com/bankbazaar/core/repository/TripUserMapRepository.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+
+import com.bankbazaar.core.model.TripUserMapEntity;
+import com.bankbazaar.core.model.TripUserCompositeKey;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+
+import java.util.List;
+
+public interface TripUserMapRepository extends JpaRepository<TripUserMapEntity, TripUserCompositeKey> {
+    @Query("SELECT t FROM TripUserMapEntity t WHERE t.userId.userId = ?1")
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/core/repository/TripUserMapRepository.java]: Dont use @Query. Use the jpa naming conventions

File: dto/pom.xml [hunk 9]
```diff
@@ -11,6 +11,15 @@
 
     <artifactId>dto</artifactId>
     <packaging>jar</packaging>
+
+    <dependencies>
+        <dependency>
+            <groupId>com.bankbazaar</groupId>
+            <artifactId>core</artifactId>
+            <version>0.0.1-SNAPSHOT</version>
+            <scope>compile</scope>
```
Comment (PinakiGhosh) [inline dto/pom.xml]: Dto should not require core

File: dto/src/main/java/com/bankbazaar/dto/model/TripActivityDto.java [hunk 10]
```diff
@@ -1,7 +1,9 @@
 package com.bankbazaar.dto.model;
+import com.bankbazaar.core.model.ActivityStatus;
```
Comment (PinakiGhosh) [inline dto/src/main/java/com/bankbazaar/dto/model/TripActivityDto.java]: I dont think this is being used anywhere

File: service/src/main/java/com/bankbazaar/service/controller/TripController.java [hunk 11]
```diff
@@ -1,61 +1,73 @@
...
-
-    @RequestMapping(method = RequestMethod.PUT)
-    public ResponseEntity<TripEntity> updateTrip(@Valid @RequestBody TripDto trip) {
-
-        TripEntity response = service.saveTrip(modelMapper.map(trip, TripEntity.class));
-        if(response == null)
+    @RequestMapping(method = RequestMethod.GET)
+    public ModelAndView viewTrip(Principal principal){
+        UserEntity userData = service.userDetails(principal).get();
+        List<TripUserMapEntity> tripList = tripUserMap.getTripsUserId(userData.getUserId());
```
Comment (PinakiGhosh) [inline service/src/main/java/com/bankbazaar/service/controller/TripController.java]: Controller should not be having all this logic. This should be in service

File: service/src/main/java/com/bankbazaar/service/controller/TripUserMapController.java [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+        model.addObject("userList", userList);
+        Optional<TripUserMapEntity> userData = manager.exists(tripId,service.userDetails(principal).get().getUserId());
+        model.addObject("role",userData.get().getUserRole().toString());
+        return model;
+    }
+    @RequestMapping(method = RequestMethod.POST)
+    public ResponseEntity<String> addUsers( @RequestParam Long tripId,@RequestParam Long userId) {
+
+        if(manager.exists(tripId,userId).isEmpty()) {
+            TripUserMapEntity tripUserMap = new TripUserMapEntity();
```
Comment (PinakiGhosh) [inline service/src/main/java/com/bankbazaar/service/controller/TripUserMapController.java]: All business logic should be in service

File: service/src/main/java/com/bankbazaar/service/manager/DataManager.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.security.Principal;
+import java.util.Optional;
+
+@Slf4j
+@Service
+public class DataManager {
```
Comment (PinakiGhosh) [inline service/src/main/java/com/bankbazaar/service/manager/DataManager.java]: Why do we need a separate manager?

**Review by PinakiGhosh — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/bankbazaar/core/manager/TripActivityEntityManager.java (hunk 2 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/core/manager/TripActivityEntityManager.java]: Actually, when i try to save "data", i get an error because all other fields in "data.tripId" is null. So, what i am doing is getting the existing "trip" from DB with "data.tripId.tripId" and then replacing the object "data.tripId" with the existing trip. This problem started when i set columns to "nullable=false".

File: core/src/main/java/com/bankbazaar/core/manager/TripActivityEntityManager.java (hunk 3 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/core/manager/TripActivityEntityManager.java]: This was unnecessary. I had implemented this right after TripUserMapping, so i followed the same structure. I have made changes to this.

File: core/src/main/java/com/bankbazaar/core/manager/TripUserMapEntityManager.java (hunk 4 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/core/manager/TripUserMapEntityManager.java]: Same like the problem above. after setting columns to nullable=false, i have to update "tripId" and "userId" foreign key objects with the existing data .

File: core/src/main/java/com/bankbazaar/core/model/TripActivityCompositeKey.java (hunk 5 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/core/model/TripActivityCompositeKey.java]: I have removed this.

File: core/src/main/java/com/bankbazaar/core/repository/TripActivityRepository.java (hunk 7 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/core/repository/TripActivityRepository.java]: This is one of the doubts i had, findAll() returns the whole table. but, i wanted to filter specific rows to my lst.

File: core/src/main/java/com/bankbazaar/core/repository/TripUserMapRepository.java (hunk 8 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/core/repository/TripUserMapRepository.java]: same as above

File: dto/pom.xml (hunk 9 above)
Comment (obedjamir) [reply dto/pom.xml]: I have removed this.

File: dto/src/main/java/com/bankbazaar/dto/model/TripActivityDto.java (hunk 10 above)
Comment (obedjamir) [reply dto/src/main/java/com/bankbazaar/dto/model/TripActivityDto.java]: this has been removed

File: core/src/main/java/com/bankbazaar/core/manager/TripActivityEntityManager.java (hunk 2 above)
Comment (PinakiGhosh) [reply core/src/main/java/com/bankbazaar/core/manager/TripActivityEntityManager.java]: But when you are inserting a TripActivity why should it try to update the Trip table? Something does not seem right there

File: service/src/main/java/com/bankbazaar/service/controller/TripUserMapController.java (hunk 12 above)
Comment (obedjamir) [reply service/src/main/java/com/bankbazaar/service/controller/TripUserMapController.java]: This has been implemented

File: service/src/main/java/com/bankbazaar/service/controller/TripController.java (hunk 11 above)
Comment (obedjamir) [reply service/src/main/java/com/bankbazaar/service/controller/TripController.java]: Changes hasve been made

File: service/src/main/java/com/bankbazaar/service/manager/DataManager.java (hunk 13 above)
Comment (obedjamir) [reply service/src/main/java/com/bankbazaar/service/manager/DataManager.java]: necessary changes made

File: core/src/main/java/com/bankbazaar/core/manager/TripActivityEntityManager.java (hunk 2 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/core/manager/TripActivityEntityManager.java]: I tried after removing that part. It is working now. I think previously there was something else that was causing the error.

Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/core/manager/TripActivityEntityManager.java]: I have found the problem. It was a problem caused by the entity class because of unproper foreign key mapping. I have made the necessary changes.

File: core/src/main/java/com/bankbazaar/core/repository/TripActivityRepository.java [hunk 14]
```diff
@@ -1,14 +1,11 @@
...
-import org.springframework.data.jpa.repository.Query;
 
 import java.util.List;
 
-public interface TripActivityRepository extends JpaRepository<TripActivityEntity, TripActivityCompositeKey> {
+public interface TripActivityRepository extends JpaRepository<TripActivityEntity, Long> {
 
-    @Query("SELECT t FROM TripActivityEntity t WHERE t.tripId.tripId = ?1")
-    public List<TripActivityEntity> getByTripId(Long tripId);
+    public List<TripActivityEntity> findAllByTripIdTripId(Long tripId);
```
Comment (PinakiGhosh) [inline core/src/main/java/com/bankbazaar/core/repository/TripActivityRepository.java]: Should this be TripIdTripId?

File: core/src/main/java/com/bankbazaar/core/model/TripActivityCompositeKey.java (hunk 6 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/core/model/TripActivityCompositeKey.java]: I tried removing @AllArgsConstructor and @NoArgsConstructor but im getting no contructor error

**Review by PinakiGhosh — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/bankbazaar/core/repository/TripActivityRepository.java (hunk 14 above)
Comment (obedjamir) [reply core/src/main/java/com/bankbazaar/core/repository/TripActivityRepository.java]: Yes, it should be TripId only.

File: service/src/main/java/com/bankbazaar/service/manager/TripActivityService.java [hunk 15]
```diff
@@ -22,40 +18,29 @@ public class TripActivityService {
...
-    private TripUserMapManager tripUserMap;
-    public ModelAndView createActivityService(Long tripId){
-        ModelAndView model = new ModelAndView("create_activity");
-        model.addObject("tripId", tripId);
-        return model;
-    }
 
-    public String addTripService(Long tripId, TripActivityDto tripActivity, Principal principal) {
+
+    public TripActivityDto addTripService(Long tripId, TripActivityDto tripActivity, Principal principal) {
```
Comment (PinakiGhosh) [inline service/src/main/java/com/bankbazaar/service/manager/TripActivityService.java]: Your service methods should not be appended with service.
Example addTripService should be addTrip

File: service/src/main/java/com/bankbazaar/service/manager/TripActivityService.java [hunk 16]
```diff
@@ -22,40 +18,29 @@ public class TripActivityService {
...
         tripActivity.setAddedBy(userService.userDetails(principal).get().getUserId());
         tripActivity.setActivityStatus("PENDING");
-        manager.saveTripActivity(modelMapper.dtoToDomain(tripActivity));
-        return "redirect:/trip-activity?tripId="+tripId;
+        TripActivityEntity response = manager.saveTripActivity(modelMapper.dtoToDomain(tripActivity));
+        return  modelMapper.domainToDto(response);
     }
 
-    public ModelAndView findTripByIdService(Long tripId,Principal principal) {
+    public List<TripActivityEntity> findTripByIdService(Long tripId,Principal principal) {
```
Comment (PinakiGhosh) [inline service/src/main/java/com/bankbazaar/service/manager/TripActivityService.java]: This should not return Entity
